### PR TITLE
Bump nixpkgs and CAF to switch to GCC11 and clang12

### PR DESCRIPTION
**Pull request by tobim**
### :notebook_with_decorative_cover: Description

This updates nixpkgs to db6e089 so we can start using GCC 11 for our static binary build. I had to integrate some fixes in the CAF submodule for that too.

### :memo: Checklist

- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Download the static binary artifacts and verify that they pass the integration tests.

### CODE REVIEW

File: nix/overlay.nix [hunk 1]
```diff
@@ -37,6 +37,10 @@ in {
     # at the optimization stage.
     # TODO: Remove when updating to CAF 0.18.
     + lib.optionalString isStatic " -std=c++17";
+    preCheck = ''
+      export LD_LIBRARY_PATH=$PWD/lib
+      export DYLD_LIBRARY_PATH=$PWD/lib
+    '';
```
Comment (dominiklohmann) [inline nix/overlay.nix:43]: Why is this necessary?

Comment (tobim) [reply nix/overlay.nix:43]: For regular builds - meaning not cross compiling - the unit tests are run after the build but before installing. Nix sets the flag `CMAKE_BUILD_WITH_INSTALL_RPATH=ON`, that means we need to add the library location here so the dynamic linker can find the `.so`s.

> But this is based on the upstream caf package which we just override with a custom version of the sources, so why isn't it taken care of by the upstream package?

Well, upstream is already at CAF 0.18, and that changed the `CMAKE_LIBRARY_OUTPUT_DIRECTORY`, so the upstream exports are different:
```
    export LD_LIBRARY_PATH=$PWD/libcaf_core:$PWD/libcaf_io
    export DYLD_LIBRARY_PATH=$PWD/libcaf_core:$PWD/libcaf_io
```
So we need to override it to work for CAF 0.17.

Comment (dominiklohmann) [reply nix/overlay.nix:43]: This is a really great reply, thanks for explaining in such detail.

**Review by dominiklohmann — APPROVED** [review]
